File: ruby/src/detectors/insufficiently-protected-credentials/insufficiently-protected-credentials.rb
=begin
 Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 SPDX-License-Identifier: Apache-2.0
=end

# {fact rule=insufficiently-protected-credentials@v1.0 defects=1}
require 'jwt'

def insufficiently_protected_credentials_noncompliant(hmac_secret)
  # Noncompliant: JWT password is hardcoded in payload.
  payload = { data: 'data', password: 12345 }
  token = JWT.encode payload, hmac_secret, 'HS256'
  puts token
end
# {/fact}

# {fact rule=insufficiently-protected-credentials@v1.0 defects=0}
def insufficiently_protected_credentials_compliant(hmac_secret)
  # Compliant: JWT password is not hardcoded.
  payload = { data: 'data', nbf: nbf }
  token = JWT.encode payload, hmac_secret, 'HS256'
  puts token
end
# {/fact}

File: jsx/src/detectors/insufficiently_protected_credentials/insufficiently_protected_credentials.js
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

// {fact rule=insufficiently-protected-credentials@v1.0 defects=1}
var express = require('express')
var loginController = express.Router()
var jwt = require("jsonwebtoken")
function insufficientlyProtectedCredentialsNoncompliant() {
    loginController.post('/', async (req, res, next) => {
        var email = req.body.email
        // Noncompliant: JWT is not signed with a strong cipher algorithm.
        var token = jwt.sign(email, process.env.JWT_SECRET, { algorithm: 'none' })
        return token
    })
}
// {/fact}


// {fact rule=insufficiently-protected-credentials@v1.0 defects=0}
var express = require('express')
var loginController = express.Router()
var jwt = require("jsonwebtoken")
function insufficientlyProtectedCredentialsCompliant() {
    loginController.post('/', async (req, res, next) => {
        var email = req.body.email
        // Compliant: JWT is signed with a strong cipher algorithm.
        var token = jwt.sign(email, process.env.JWT_SECRET, { algorithm: 'RS256' })
        return token
    })
}
// {/fact}